pkg/services/datasources/service/secrets_mig.go

```diff
@@ -0,0 +1,92 @@
+package service
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/infra/kvstore"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+)
+
+const (
+	secretMigrationStatusKey       = "secretMigrationStatus"
+	compatibleSecretMigrationValue = "compatible"
+	completeSecretMigrationValue   = "complete"
+)
+
+type DataSourceSecretMigrationService struct {
+	dataSourcesService datasources.DataSourceService
+	kvStore            *kvstore.NamespacedKVStore
+	features           featuremgmt.FeatureToggles
+}
+
+func ProvideDataSourceMigrationService(
+	dataSourcesService datasources.DataSourceService,
+	kvStore kvstore.KVStore,
+	features featuremgmt.FeatureToggles,
+) *DataSourceSecretMigrationService {
+	return &DataSourceSecretMigrationService{
+		dataSourcesService: dataSourcesService,
+		kvStore:            kvstore.WithNamespace(kvStore, 0, secretType),
+		features:           features,
+	}
+}
+
+func (s *DataSourceSecretMigrationService) Migrate(ctx context.Context) error {
+	migrationStatus, _, err := s.kvStore.Get(ctx, secretMigrationStatusKey)
+	if err != nil {
+		return err
+	}
+
+	disableSecretsCompatibility := s.features.IsEnabled(featuremgmt.FlagDisableSecretsCompatibility)
+	needCompatibility := migrationStatus != compatibleSecretMigrationValue && !disableSecretsCompatibility
+	needMigration := migrationStatus != completeSecretMigrationValue && disableSecretsCompatibility
+
+	if needCompatibility || needMigration {
```

Comment:
Here we check the value of the secretMigrationStatus key retrieved from the kvStore with the disableSecretCompatibility flag to tell if a migration is required.

- If migrationStatus != “compatible” && disableSecretCompatibility == false
   - We need migration for 2 reasons:
      - Migrate secrets from legacy to unified
      - Or from unified back to legacy
- If migrationStatus != “complete” && disableSecretCompatibility == true
   - Again we need migration for 2 reasons:
      - Migrate secrets from legacy to unified
      - And delete secrets from legacy

On both cases the UpdateDataSource function is the one that does the migration based on how the disableSecretCompatibility flag is set.

In case we already migrated secrets from legacy to unified with the flag set to one value, if that value changes, we need to re-run migration again for the second reason on both cases.